File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
'''Create Rectangle class'''


class Rectangle:
    '''Defining class attributes'''
    def __init__(self, width=0, height=0):
        self.__width = width
        self.__height = height

    @property
    def width(self):
        '''Retrieve width attribute'''
        return self.__width

    @width.setter
    def width(self, value):
        '''Sets width value and ensures width is an integer'''
        if type(value) != int:
            raise TypeError("width must be an integer")
        if value < 0:
            raise ValueError("width must be >= 0")
        self.__width = value

    @property
    def height(self):
        '''Retrieve height attribute'''
        return self.__height

    @height.setter
    def height(self, value):
        '''Sets height and ensures it is an integer not less than 0'''
        if type(value) != int:
            raise TypeError("height must be an integer")
        if value < 0:
            raise ValueError("height must be >= 0")
        self.__height = value

    def area(self):
        '''Returns value of rectangle area'''
        self.height = self.__height
        self.width = self.__width
        return self.height * self.width

    def perimeter(self):
        '''Returns perimeter of rectangle object'''
        if self.__height == 0 or self.__width == 0:
            return 0
        self.height = self.__height
        self.width = self.__width
        return 2 * (self.height + self.width)

    def __str__(self):
        '''Runs when print() or str() is used on a Rectangle objects'''
        self.height = self.__height
        self.width = self.__width
        for i in range(self.height):
            for j in range(self.width):
                print("#", end="")
            print()
        return str()

    def __repr__(self):
        '''Runs when repr() is called'''
        return f"Rectangle"

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node
class Rectangle {
  constructor (w, h) {
    if (w && h) {
      if ((w > 0) && (h > 0)) {
        this.width = w;
        this.height = h;
      }
    }
  }

  print () {
    let h;
    let j;
    for (h = 0; h < this.height; h++) {
      let shapeString = '';
      for (j = 0; j < this.width; j++) {
        shapeString += 'X';
      }
      console.log(shapeString);
    }
  }

  rotate () {
    const middleMan = this.width;
    this.width = this.height;
    this.height = middleMan;
  }

  double () {
    this.width *= 2;
    this.height *= 2;
  }
}

module.exports = Rectangle;
